includes/gateways/paypal-standard.php

```diff
@@ -1122,105 +1124,119 @@ function edd_paypal_link_transaction_id( $transaction_id, $payment_id ) {
+ * @param int $refund_id ID of the newly created refund record.
+ * @param bool $all_refunded Whether or not this was a full refund.
  *
- * @param object|EDD_Payment $payment The current payment ID.
- * @return void
+ * @since 3.0
  */
-function edd_maybe_refund_paypal_purchase( EDD_Payment $payment ) {
-
-
-	if ( ! current_user_can( 'edit_shop_payments', $payment->ID ) ) {
+function edd_paypal_maybe_refund_transaction( $order_id, $refund_id, $all_refunded ) {
+	if ( ! current_user_can( 'edit_shop_payments', $order_id ) ) {
 		return;
 	}
 
-	if ( empty( $_POST['edd-paypal-refund'] ) ) {
+	if ( empty( $_POST['data'] ) ) {
 		return;
 	}
 
-	$processed = $payment->get_meta( '_edd_paypal_refunded', true );
-
-	// If the status is not set to "refunded", return early.
-	if ( 'complete' !== $payment->old_status && 'revoked' !== $payment->old_status ) {
+	$order = edd_get_order( $order_id );
+	if ( empty( $order->gateway ) || 'paypal' !== $order->gateway ) {
 		return;
 	}
 
-	// If not PayPal/PayPal Express, return early.
-	if ( 'paypal' !== $payment->gateway ) {
+	// Get our data out of the serialized string.
+	parse_str( $_POST['data'], $form_data );
+
+	if ( empty( $form_data['edd-paypal-refund'] ) ) {
+		edd_add_note( array(
+			'object_id'   => $order_id,
+			'object_type' => 'order',
+			'user_id'     => is_admin() ? get_current_user_id() : 0,
+			'content'     => __( 'Transaction not refunded in PayPal, as checkbox was not selected.', 'easy-digital-downloads' )
+		) );
+
 		return;
 	}
 
-	// If the payment has already been refunded in the past, return early.
-	if ( $processed ) {
+	$refund = edd_get_order( $refund_id );
+	if ( empty( $refund->total ) ) {
 		return;
 	}
 
-	// Process the refund in PayPal.
-	edd_refund_paypal_purchase( $payment );
-
+	edd_refund_paypal_purchase( $order, $refund );
 }
-add_action( 'edd_pre_refund_payment', 'edd_maybe_refund_paypal_purchase', 999 );
+add_action( 'edd_refund_order', 'edd_paypal_maybe_refund_transaction', 10, 3 );
 
 /**
  * Refunds a purchase made via PayPal.
  *
  * @since  2.6.0
  *
- * @param object|int $payment The payment ID or object to refund.
+ * @param EDD_Payment|Order|int $payment_id_or_object The ID or object of the order to refund.
+ * @param Order|null            $refund_object        Optional. The refund object associated with this
+ *                                                    transaction refund. If provided, then the refund
+ *                                                    amount is used as the transaction refund amount (used for
+ *                                                    partial refunds), and an EDD transaction record will be
+ *                                                    inserted.
+ *
  * @return void
  */
-function edd_refund_paypal_purchase( $payment ) {
+function edd_refund_paypal_purchase( $payment_id_or_object, $refund_object = null ) {
+	/*
+	 * Internally we want to work with an Order object, but we also need
```

Comment:
This comment indicates that the function will be working with the order object, but it's still using the `$payment` object in the `$credentials` array and for the transaction ID (line 1258). Is that intentional?